ksql-api/src/main/java/io/confluent/ksql/api/server/ReactiveSubscriber.java

```diff
@@ -0,0 +1,176 @@
+/*
+ * Copyright 2020 Confluent Inc.
+ *
+ * Licensed under the Confluent Community License (the "License"); you may not use
+ * this file except in compliance with the License.  You may obtain a copy of the
+ * License at
+ *
+ * http://www.confluent.io/confluent-community-license
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OF ANY KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations under the License.
+ */
+
+package io.confluent.ksql.api.server;
+
+import io.vertx.core.Context;
+import io.vertx.core.Vertx;
+import java.util.Objects;
+import org.reactivestreams.Subscriber;
+import org.reactivestreams.Subscription;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+/**
+ * A reactive streams subscriber which handles much of the plumbing for you. Override {@link
+ * #afterSubscribe}, {@link #handleValue}, {@link #handleComplete} and {@link #handleError} to
+ * create your specific implementation.
+ *
+ * @param <T> The type of the value
+ */
+public class ReactiveSubscriber<T> implements Subscriber<T> {
+
+  private static final Logger log = LoggerFactory.getLogger(ReactiveSubscriber.class);
+
+  protected final Context context;
+  private Subscription subscription;
+  private boolean complete;
+
+  /**
+   * Construct a ReactiveSubscriber
+   *
+   * @param context The Vert.x context to use for the subscriber - the subscriber code will always
+   *                be executed on this context. This ensures the code is never executed
+   *                concurrently by more than one thread.
+   */
+  public ReactiveSubscriber(final Context context) {
+    this.context = context;
+  }
+
+  @Override
+  public void onSubscribe(final Subscription s) {
+    Objects.requireNonNull(s);
+    context.runOnContext(v -> doOnSubscribe(s));
+  }
+
+  @Override
+  public void onNext(final T t) {
+    Objects.requireNonNull(t);
+    context.runOnContext(v -> doOnNext(t));
+  }
+
+  @Override
+  public void onError(final Throwable t) {
+    Objects.requireNonNull(t);
+    context.runOnContext(v -> doOnError(t));
+  }
+
+  @Override
+  public void onComplete() {
+    context.runOnContext(v -> doOnComplete());
+  }
+
+  protected void afterSubscribe(final Subscription subscription) {
+  }
+
+  protected void handleValue(final T t) {
+  }
+
+  protected void handleComplete() {
+  }
+
+  protected void handleError(final Throwable t) {
+  }
+
+  private void doOnSubscribe(final Subscription subscription) {
+    checkContext();
```

Comment:
I think it's better to keep the check inside the method as is, in case that method accidentally gets called from elsewhere that's on the wrong context in a future refactoring.